docs/slotfilling.rst

```diff
@@ -3,12 +3,284 @@
+                Empty dict is converted to a mapping of
+                the slot to the extracted entity with the same name
+            """
+
+            return {}
+
+
+        def validate(self, dispatcher, tracker, domain):
+              # type: (CollectingDispatcher, Tracker, Dict[Text, Any]) -> List[Dict]
+              """Validate extracted value of requested slot
+                  else reject execution of the form action
+                  Subclass this method to add custom validation and rejection logic
+              """
+              # extract other slots that were not requested
+              # but set by corresponding entity
+              slot_values = self.extract_other_slots(dispatcher, tracker, domain)
+
+              # extract requested slot
+              slot_to_fill = tracker.get_slot(REQUESTED_SLOT)
+              if slot_to_fill:
+                  slot_values.update(self.extract_requested_slot(dispatcher,
+                                                                 tracker, domain))
+                  if not slot_values:
+                      # reject to execute the form action
+                      # if some slot was requested but nothing was extracted
+                      # it will allow other policies to predict another action
+                      raise ActionExecutionRejection(self.name(),
+                                                     "Failed to validate slot {0} "
+                                                     "with action {1}"
+                                                     "".format(slot_to_fill,
+                                                               self.name()))
+
+              # validation succeed, set slots to extracted values
+              return [SlotSet(slot, value) for slot, value in slot_values.items()]
+
+          def submit(self, dispatcher, tracker, domain):
+              # type: (CollectingDispatcher, Tracker, Dict[Text, Any]) -> List[Dict]
+              """Define what the form has to do
+                  after all required slots are filled"""
+
+              raise NotImplementedError("A form must implement a submit method")
+
+
+For the RestaurantForm, you first define it's name and the list of required slots:
+
+.. code-block:: python
+
+ class RestaurantForm(FormAction):
+    """Example of a custom form action"""
+
+    def name(self):
+        # type: () -> Text
+        """Unique identifier of the form"""
+
+        return "restaurant_form"
+
+    @staticmethod
+    def required_slots():
+        # type: () -> List[Text]
+        """A list of required slots that the form has to fill"""
+
+        return ["cuisine", "num_people", "outdoor_seating",
+                "preferences", "feedback"]
+
+
+The way this works is that every time you call this action, it will pick one
+slot from the ``required_slots()`` that's still missing and ask the user for it.
+This is done by using the ``utter_ask_{slot_name}`` template that you need to
```

Comment:
Probably best to talk about `request_next_slot()` and custom logic for specific utterances here

Reply:
hmm not sure. what kind of situations do you mean? I think we decided that we wanted to advise people to implement custom logic in `required_slots()`